wagtail/wagtailcore/models.py

```diff
@@ -40,8 +39,86 @@
 
 PAGE_TEMPLATE_VAR = 'page'
 
+SITE_CACHE = {}
+REQUEST_SITE_MATCH_CACHE = {}
+
 
 class SiteManager(models.Manager):
+
+    @staticmethod
+    def get_hostname_and_port_from_request(request):
+        """Reliably extract 'hostname' and 'port' values from the supplied
+        ``HttpRequest`` instance (without raising exceptions), and return them
+        as a tuple.
+        """
+        try:
+            hostname = request.get_host().split(':')[0]
+        except KeyError:
+            hostname = None
+        try:
+            port = request.get_port()
+        except KeyError:
+            port = None
+        return hostname, port
+
+    @staticmethod
+    def clear_site_cache():
+        global SITE_CACHE
+        global REQUEST_SITE_MATCH_CACHE
+        SITE_CACHE = {}
+        REQUEST_SITE_MATCH_CACHE = {}
+
+    def _get_by_id(self, site_id):
+        if site_id not in SITE_CACHE:
+            site = self.get(pk=site_id)
+            SITE_CACHE[site_id] = site
+        return SITE_CACHE[site_id]
+
+    def _cache_and_return_match(self, site, match_cache_key):
+        SITE_CACHE[site.pk] = site
+        REQUEST_SITE_MATCH_CACHE[match_cache_key] = site.pk
+        return SITE_CACHE[site.pk]
+
+    def get_default(self):
+        """Returns the 'default' ``Site`` or raise an exception in no site is
+        set as the default. The result is cached."""
+        key = 'default'
+        if key in REQUEST_SITE_MATCH_CACHE:
+            return self._get_by_id(REQUEST_SITE_MATCH_CACHE['default'])
+
+        match = self.get(is_default_site=True)
+        return self._cache_and_return_match(match, key)
+
+    def get_for_request(self, request):
+        """Return the site responsible for dealing with the supplied
+        ``HttpRequest`` instance. The result is cached."""
+        hostname, port = self.get_hostname_and_port_from_request(request)
+        key = "%s:%s" % (hostname, port)
+
+        if key in REQUEST_SITE_MATCH_CACHE:
+            return self._get_by_id(REQUEST_SITE_MATCH_CACHE[key])
+
+        if hostname:
+            if port:
+                try:
+                    # Try to find a site matching both hostname and port
+                    match = self.get(hostname=hostname, port=port)
+                    return self._cache_and_return_match(match, key)
+                except Site.DoesNotExist:
+                    pass
+
+            try:
+                # If there's only one site matching the hostname, use that,
+                # since there's no ambiguity
+                match = self.get(hostname=hostname)
+                return self._cache_and_return_match(match, key)
+            except(Site.DoesNotExist, Site.MultipleObjectsReturned):
+                pass
+
+        # Fall back to using default site
+        match = self.get_default()
+        return self._cache_and_return_match(match, key)
```

Comment:
In all this method, `match` are `Site` instances. This is a bit misleading when we read what was in `get_site_for_hostname`, which was a very different type of match. Prefer `site`.

Reply:
Yes, I can see your point there. Using `site` consistently would be less ambiguous.